### PR TITLE
Bookstore: connect to API

### PR DESCRIPTION
**Pull request by techmoves**
in this repo i did this:

- [x]  Make use of the [Bookstore API](https://www.notion.so/Bookstore-API-51ea269061f849118c65c0a53e88a739) as your backend.
    

- [x] Use [axios](https://axios-http.com/docs/intro) with [createAsyncThunk](https://redux-toolkit.js.org/api/createAsyncThunk) to make your HTTP requests to the backend

    

- [x] Fetch your books from the Bookstore API after the component has rendered

   

- [x]  Refactoed your add book and remove book features to persist your changes in the server

### CODE REVIEW

File: src/components/BooksList.js [hunk 1]
```diff
@@ -1,14 +1,29 @@
...
+  useEffect(() => {
+    dispatch(getBooks());
+  }, [dispatch]);
+
+  const itemComponents = Object.keys(booksItem).flatMap((key) => booksItem[key].map((book) => (
+    <BooksItem
+      key={`${key}-${book.title}`}
+      list={{ id: key, ...book }}
+    />
+  )));
```
Comment (topeogunleye) [inline src/components/BooksList.js]: Hi, you have a good start with your code, but there is an error that you need to fix. Your `booksItem` variable is either undefined or null, and therefore cannot be used with the `Object.keys` method. This causes an error message: Cannot convert undefined or null to object. You need to make sure that the `booksItem` variable is always an object with some keys and values. You can also use a conditional operator or a default value to handle the case when the `booksItem` variable is undefined or null. For example:

```
// Using a conditional operator
const itemComponents = booksItem ? Object.keys(booksItem).flatMap((key) => booksItem[key].map((book) => (
    <BooksItem
      key={`${key}-${book.title}`}
      list={{ id: key, ...book }}
    />
  ))) : [];

// Using a default value
const itemComponents = Object.keys(booksItem || {}).flatMap((key) => booksItem[key].map((book) => (
    <BooksItem
      key={`${key}-${book.title}`}
      list={{ id: key, ...book }}
    />
  )));

```
Please make this change and resubmit your code for review. If you have any questions, feel free to ask. Thank you.

**Review by topeogunleye — CHANGES_REQUESTED** [review]
Hi @techmoves, 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.


 ## Highlights
- You have imported createAsyncThunk from "@reduxjs/toolkit" and used it to create a fetchBooks action creator that makes a GET request to the API using axios and returns the response data. 🙌
- You have added an extraReducers property to your books slice and handled the pending, fulfilled and rejected cases for the fetchBooks action. This ensures that your state reflects the status and result of the async action. 🙌
- You have used useEffect to dispatch the fetchBooks action in your App component. This ensures that your component fetches the data from the API when it mounts and renders the books accordingly. 🙌
- You have refactored your addBook and removeBook action creators to use createAsyncThunk as well. They make POST and DELETE requests to the API using axios and return a promise. They also automatically dispatch actions that add or remove a book from the state. 🙌

 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

You can also consider:

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag @topeogunleye in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: src/components/BooksList.js (hunk 1 above)
Comment (techmoves) [reply src/components/BooksList.js]: Thanks alot it help alot

**Review by emmanuellekamwa — APPROVED** [review]
**STATUS: APPROVED ✔🍾🎉**
Hello @techmoves 👋
Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## Highlights

- Descriptive PR.
- Good commit messages.
- Correct git-flow.

Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: src/components/BooksList.js (hunk 1 above)
Comment (topeogunleye) [reply src/components/BooksList.js]: You're welcome. I'm glad it helped. Keep up the good work!

**Pull request merged**